### PR TITLE
[CDRIVER-6017] BSON Validation Refactor (#2026) (Cherry-pick for 1.30.x)

### PR DESCRIPTION
**Pull request by vector-of-bool**
This is a cherry-pick of changes from #2026 for the 1.30 release.

Because the new code makes use of files in the `mlib/` directory, those files have been copied from `master` into this PR, as well as a `.clang-format` file to keep their formatting since the formatting has diverged since. The code in this directory is not used anywhere except as part of the new validation code and its relevant test cases.

The original commit message is retained below and should be included in the final merge.

---

* New BSON validation routine rewrite

The new `bson_validate` implementation does not
make use of the error-prone `bson_visit` APIs. Instead, it is written as a simple recursive validator. The new validator respects requests for UTF-8 validation properly.

* Stop validating at 1000 depth, preventing stack overflow
* Replace most BSON validation tests with generated ones

The existing test cases used BSON files, and didn't have any commentary on what they were actually testing. New test cases are generated from a Python shorthand and contain the tested bytes inline, with a distinct test case for each actual validation scenario.

* Disable UTF-8 validation by default on CRUD APIs
* Document and tweak the value of BSON_VALIDATE_CORRUPT
* Add test cases related to the overlong null encoding
* Tweak JS scope validation to permit more obj keys
* Add a NEWS entry for validation changes.
* Allow `-private.h` headers to not include the prelude header

---------

### CODE REVIEW

File: src/libbson/src/bson/validate.c [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bson/validate-private.h>
+#include <mlib/intencode.h>
```
Comment (kevinAlbs) [inline src/libbson/src/bson/validate.c]: At the risk of having diverging changes: can mlib be removed to reduce the size of the backport? The release branches are intended to be stable. I am mildly concerned mlib could introduce some unexpected compile change.

If tests are removed I expect all that is needed is to replace `mlib_read_u32le` with `   BSON_UINT32_FROM_LE` and `mlib_check` with `BSON_ASSERT`.

File: src/libbson/NEWS [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+=============
+
+Fixes:
+
+  * Fix CMake configure error on macOS with CMake 4.
+
+Thanks to everyone who contributed to the development of this release.
+
+  * Kevin Albertson
+
```
Comment (eramongodb) [inline src/libbson/NEWS]:
```suggestion
```
Stray entries specific to 2.0.1?

File: src/libbson/tests/validate-tests.py [hunk 3]
```diff
@@ -0,0 +1,1479 @@
...
+            subsequent_indent=" " * 4,
+            initial_indent=" " * 4,
+            width=80,
+        )
+    )
+    yield "\n  };\n"
+    yield from [
+        # Initialize a BSON doc that points to the byte array
+        "  bson_t doc;\n",
+        "  BSON_ASSERT(bson_init_static(&doc, bytes, sizeof bytes));\n",
```
Comment (eramongodb) [inline src/libbson/tests/validate-tests.py]:
```suggestion
        "  ASSERT(bson_init_static(&doc, bytes, sizeof bytes));\n",
```
Prefer test suite assertions.

File: src/libbson/tests/validate-tests.py [hunk 4]
```diff
@@ -0,0 +1,1479 @@
...
+        "  size_t offset = 999999;\n"
+        # Do the actual validation:
+        f"  const bool is_valid = bson_validate_with_error_and_offset(&doc, {case.flags}, &offset, &error);\n",
+    ]
+    is_error = case.error.code != "0"
+    yield from [
+        "  BSON_ASSERT(!is_valid);\n" if is_error else "  ASSERT_OR_PRINT(is_valid, error);\n",
+        f"  BSON_ASSERT(error.code == {case.error.code});\n",
+        f"  BSON_ASSERT(!strcmp(error.message, {json.dumps(case.error.message)}));\n",
+        f"  BSON_ASSERT(offset == {case.error.offset});\n" if is_error else "",
```
Comment (eramongodb) [inline src/libbson/tests/validate-tests.py]:
```suggestion
        "  ASSERT(!is_valid);\n" if is_error else "  ASSERT_OR_PRINT(is_valid, error);\n",
        f"  ASSERT_CMPINT32(error.code, ==, {case.error.code});\n",
        f"  ASSERT_CMPSTR(error.message, {json.dumps(case.error.message)});\n",
        f"  ASSERT_CMPSIZE_T(offset, ==, {case.error.offset});\n" if is_error else "",
```
Prefer test suite assertions.

File: src/libbson/NEWS (hunk 2 above)
Comment (vector-of-bool) [reply src/libbson/NEWS]: ~~I believe this waits for the release process, maybe? @kevinAlbs~~

I misread. Plz ignore

**Comment by faridguitouni00-debug** [comment]
Faridguitouni00@gmail.com

**Comment by faridguitouni00-debug** [comment]
changements programmation YouTube profils il limitée provisionnelle avec IA complet dernier 2025
